File: bindings/go/kryptic.go
// Kryptic Go client
//
// Communicates with a running Kryptic server (python -m kryptic serve).
// Zero external dependencies — uses only the standard library.
//
// Run the example:
//   go run kryptic.go
package main

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

// ── client ─────────────────────────────────────────────────────────────────

type KrypticClient struct {
	Base   string
	client *http.Client
}

func NewKrypticClient(host string, port int) *KrypticClient {
	return &KrypticClient{
		Base: fmt.Sprintf("http://%s:%d", host, port),
		client: &http.Client{
			Timeout: 60 * time.Second,
		},
	}
}

func New() *KrypticClient {
	return NewKrypticClient("127.0.0.1", 7890)
}

func (k *KrypticClient) request(method, path string, body any) (map[string]any, error) {
	var reqBody io.Reader
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		reqBody = bytes.NewReader(b)
	}

	req, err := http.NewRequest(method, k.Base+path, reqBody)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := k.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	if ok, _ := result["ok"].(bool); !ok {
		errMsg, _ := result["error"].(string)
		return nil, fmt.Errorf("kryptic error: %s", errMsg)
	}
	return result, nil
}

func (k *KrypticClient) Health() (map[string]any, error) {
	return k.request("GET", "/health", nil)
}

func (k *KrypticClient) Session() (*KrypticSession, error) {
	r, err := k.request("POST", "/sessions", map[string]any{})
	if err != nil {
		return nil, err
	}
	id, _ := r["session_id"].(string)
	return &KrypticSession{client: k, ID: id}, nil
}

func (k *KrypticClient) HTTPGet(url string, headers map[string]string) (map[string]any, error) {
	return k.request("POST", "/http/get", map[string]any{"url": url, "headers": headers})
}

func (k *KrypticClient) HTTPPost(url string, jsonBody any) (map[string]any, error) {
	return k.request("POST", "/http/post", map[string]any{"url": url, "json": jsonBody})
}

func (k *KrypticClient) HTTPBatch(urls []string) ([]any, error) {
	r, err := k.request("POST", "/http/batch", map[string]any{"urls": urls})
	if err != nil {
		return nil, err
	}
	results, _ := r["results"].([]any)
	return results, nil
}

// ── session ─────────────────────────────────────────────────────────────────

type KrypticSession struct {
	client *KrypticClient
	ID     string
}

func (s *KrypticSession) post(action string, body any) (map[string]any, error) {
	return s.client.request("POST", "/sessions/"+s.ID+"/"+action, body)
}

func (s *KrypticSession) get(action string) (map[string]any, error) {
	return s.client.request("GET", "/sessions/"+s.ID+"/"+action, nil)
}

func (s *KrypticSession) Goto(url string) error {
	_, err := s.post("goto", map[string]any{"url": url, "wait_until": "domcontentloaded"})
	return err
}

func (s *KrypticSession) Title() (string, error) {
	r, err := s.get("title")
	if err != nil {
		return "", err
	}
	return r["title"].(string), nil
}

func (s *KrypticSession) HTML() (string, error) {
	r, err := s.get("html")
	if err != nil {
		return "", err
	}
	return r["html"].(string), nil
}

func (s *KrypticSession) URL() (string, error) {
	r, err := s.get("url")
	if err != nil {
		return "", err
	}
	return r["url"].(string), nil
}

func (s *KrypticSession) Text(selector string) (string, error) {
	r, err := s.post("text", map[string]any{"selector": selector})
	if err != nil {
		return "", err
	}
	return r["text"].(string), nil
}

func (s *KrypticSession) Click(selector string) error {
	_, err := s.post("click", map[string]any{"selector": selector})
	return err
}

func (s *KrypticSession) Fill(selector, value string) error {
	_, err := s.post("fill", map[string]any{"selector": selector, "value": value})
	return err
}

func (s *KrypticSession) Evaluate(js string) (any, error) {
	r, err := s.post("evaluate", map[string]any{"js": js})
	if err != nil {
		return nil, err
	}
	return r["result"], nil
}

func (s *KrypticSession) Screenshot(fullPage bool) ([]byte, error) {
	r, err := s.post("screenshot", map[string]any{"full_page": fullPage})
	if err != nil {
		return nil, err
	}
	data, _ := r["data"].(string)
	return base64.StdEncoding.DecodeString(data)
}

func (s *KrypticSession) BlockResources(types []string) error {
	_, err := s.post("block", map[string]any{"resource_types": types})
	return err
}

func (s *KrypticSession) WaitFor(selector, state string) error {
	_, err := s.post("wait_for", map[string]any{"selector": selector, "state": state})
	return err
}

func (s *KrypticSession) Close() error {
	_, err := s.client.request("DELETE", "/sessions/"+s.ID, nil)
	return err
}

// ── example main ─────────────────────────────────────────────────────────────

func main() {
	k := New()

	health, err := k.Health()
	if err != nil {
		fmt.Fprintln(os.Stderr, "Cannot reach Kryptic server:", err)
		fmt.Fprintln(os.Stderr, "Start it with: python -m kryptic serve")
		os.Exit(1)
	}
	fmt.Println("Server health:", health)

	s, err := k.Session()
	if err != nil {
		panic(err)
	}
	defer s.Close()

	_ = s.BlockResources([]string{"image", "stylesheet", "font", "media"})
	if err := s.Goto("https://example.com"); err != nil {
		panic(err)
	}

	title, _ := s.Title()
	h1, _ := s.Text("h1")
	fmt.Println("Title:", title)
	fmt.Println("H1:", h1)

	resp, _ := k.HTTPGet("https://httpbin.org/get", nil)
	fmt.Println("HTTP status:", resp["status"])

	urls := []string{"https://example.com", "https://example.org", "https://iana.org"}
	batch, _ := k.HTTPBatch(urls)
	for _, r := range batch {
		rm := r.(map[string]any)
		u := strings.Split(fmt.Sprintf("%v", rm["url"]), "?")[0]
		fmt.Printf("  %.0f  %s\n", rm["status"], u)
	}
}

File: bindings/rust/kryptic.rs
/*!
Kryptic Rust client

Communicates with a running Kryptic server (python -m kryptic serve).

Dependencies (add to Cargo.toml):
    [dependencies]
    ureq = { version = "2", features = ["json"] }
    serde = { version = "1", features = ["derive"] }
    serde_json = "1"
    base64 = "0.22"

Run the example:
    cargo run
*/

use base64::{engine::general_purpose::STANDARD, Engine};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;

// ── Client ────────────────────────────────────────────────────────────────────

pub struct KrypticClient {
    base: String,
}

impl KrypticClient {
    pub fn new(host: &str, port: u16) -> Self {
        Self {
            base: format!("http://{}:{}", host, port),
        }
    }

    pub fn default() -> Self {
        Self::new("127.0.0.1", 7890)
    }

    fn post(&self, path: &str, body: Value) -> Result<Value, Box<dyn std::error::Error>> {
        let url = format!("{}{}", self.base, path);
        let resp: Value = ureq::post(&url).send_json(body)?.into_json()?;
        if resp["ok"] != true {
            return Err(resp["error"].as_str().unwrap_or("unknown error").into());
        }
        Ok(resp)
    }

    fn get(&self, path: &str) -> Result<Value, Box<dyn std::error::Error>> {
        let url = format!("{}{}", self.base, path);
        let resp: Value = ureq::get(&url).call()?.into_json()?;
        Ok(resp)
    }

    fn delete(&self, path: &str) -> Result<Value, Box<dyn std::error::Error>> {
        let url = format!("{}{}", self.base, path);
        let resp: Value = ureq::delete(&url).call()?.into_json()?;
        Ok(resp)
    }

    pub fn health(&self) -> Result<Value, Box<dyn std::error::Error>> {
        self.get("/health")
    }

    pub fn session(&self) -> Result<KrypticSession, Box<dyn std::error::Error>> {
        let resp = self.post("/sessions", json!({}))?;
        let id = resp["session_id"]
            .as_str()
            .ok_or("no session_id")?
            .to_string();
        Ok(KrypticSession { client: self, id })
    }

    pub fn http_get(
        &self,
        url: &str,
        headers: Option<HashMap<String, String>>,
    ) -> Result<Value, Box<dyn std::error::Error>> {
        self.post("/http/get", json!({ "url": url, "headers": headers }))
    }

    pub fn http_post(
        &self,
        url: &str,
        json_body: Value,
    ) -> Result<Value, Box<dyn std::error::Error>> {
        self.post("/http/post", json!({ "url": url, "json": json_body }))
    }

    pub fn http_batch(
        &self,
        urls: &[&str],
    ) -> Result<Vec<Value>, Box<dyn std::error::Error>> {
        let resp = self.post("/http/batch", json!({ "urls": urls }))?;
        Ok(resp["results"].as_array().cloned().unwrap_or_default())
    }
}

// ── Session ───────────────────────────────────────────────────────────────────

pub struct KrypticSession<'a> {
    client: &'a KrypticClient,
    pub id: String,
}

impl<'a> KrypticSession<'a> {
    fn post_action(
        &self,
        action: &str,
        body: Value,
    ) -> Result<Value, Box<dyn std::error::Error>> {
        self.client
            .post(&format!("/sessions/{}/{}", self.id, action), body)
    }

    fn get_action(&self, action: &str) -> Result<Value, Box<dyn std::error::Error>> {
        self.client.get(&format!("/sessions/{}/{}", self.id, action))
    }

    pub fn goto(&self, url: &str) -> Result<(), Box<dyn std::error::Error>> {
        self.post_action(
            "goto",
            json!({ "url": url, "wait_until": "domcontentloaded" }),
        )?;
        Ok(())
    }

    pub fn title(&self) -> Result<String, Box<dyn std::error::Error>> {
        let r = self.get_action("title")?;
        Ok(r["title"].as_str().unwrap_or("").to_string())
    }

    pub fn html(&self) -> Result<String, Box<dyn std::error::Error>> {
        let r = self.get_action("html")?;
        Ok(r["html"].as_str().unwrap_or("").to_string())
    }

    pub fn url(&self) -> Result<String, Box<dyn std::error::Error>> {
        let r = self.get_action("url")?;
        Ok(r["url"].as_str().unwrap_or("").to_string())
    }

    pub fn text(&self, selector: &str) -> Result<String, Box<dyn std::error::Error>> {
        let r = self.post_action("text", json!({ "selector": selector }))?;
        Ok(r["text"].as_str().unwrap_or("").to_string())
    }

    pub fn click(&self, selector: &str) -> Result<(), Box<dyn std::error::Error>> {
        self.post_action("click", json!({ "selector": selector }))?;
        Ok(())
    }

    pub fn fill(&self, selector: &str, value: &str) -> Result<(), Box<dyn std::error::Error>> {
        self.post_action("fill", json!({ "selector": selector, "value": value }))?;
        Ok(())
    }

    pub fn evaluate(&self, js: &str) -> Result<Value, Box<dyn std::error::Error>> {
        let r = self.post_action("evaluate", json!({ "js": js }))?;
        Ok(r["result"].clone())
    }

    pub fn screenshot(&self, full_page: bool) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
        let r = self.post_action("screenshot", json!({ "full_page": full_page }))?;
        let data = r["data"].as_str().unwrap_or("");
        Ok(STANDARD.decode(data)?)
    }

    pub fn block_resources(
        &self,
        types: &[&str],
    ) -> Result<(), Box<dyn std::error::Error>> {
        self.post_action("block", json!({ "resource_types": types }))?;
        Ok(())
    }

    pub fn wait_for(
        &self,
        selector: &str,
        state: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        self.post_action("wait_for", json!({ "selector": selector, "state": state }))?;
        Ok(())
    }

    pub fn close(&self) -> Result<(), Box<dyn std::error::Error>> {
        self.client
            .delete(&format!("/sessions/{}", self.id))?;
        Ok(())
    }
}

// ── Example main ─────────────────────────────────────────────────────────────

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let k = KrypticClient::default();

    let health = k.health()?;
    println!("Server: {}", health);

    let s = k.session()?;
    s.block_resources(&["image", "stylesheet", "font", "media"])?;
    s.goto("https://example.com")?;
    println!("Title: {}", s.title()?);
    println!("H1: {}", s.text("h1")?);
    s.close()?;

    let resp = k.http_get("https://httpbin.org/get", None)?;
    println!("HTTP status: {}", resp["status"]);

    let batch = k.http_batch(&["https://example.com", "https://example.org"])?;
    for r in &batch {
        println!("  {}  {}", r["status"], r["url"].as_str().unwrap_or(""));
    }

    Ok(())
}
